### ISSUE TITLE
Way to run tests in parallel without wdio becoming a memory hog

### ISSUE THREAD

**Issue by alfonso-presa** [webdriverio/webdriverio#4235]
**Is your feature request related to a problem? Please describe.**
We are running test against a cloud provided grid and we have enough instances to run several tests in parallel (want to reach ~100 browsers). The problem is that wdio spawns different processes with a memory footprint of around 100MB each, making it unscalable to that point inside our automation pipeline (will require around 10GB memory).

**Describe the solution you'd like**
I would like a way to use the webdriver.io runner that has the possibility of reusing a process for several browsers so that the cost of memory of each browser used in the pipeline. It would not necessarily mean that we are using only one process (as only one thread at a time might lead to a CPU bottleneck), but it would be nice if we could tune the amount of browsers managed by each process.

**Describe alternatives you've considered**
The only alternative that comes to my mind is having each process using somehow the [VM](https://nodejs.org/api/vm.html) capability of nodejs to bootstrap different contexts so that multiple browsers can be managed at a time. I've not look into this in so much depth, but it looks like doable just by looking at the API description.

**Comment by mgrybyk** [comment]
@alfonso-presa can you please describe your environment:
- wdio versions 
- wdio conf (capabilities, max instances, etc)
- etc

**Comment by alfonso-presa** [comment]
@mgrybyk Sure. It wasn't requested in the feature request template though :-).

Memory measurements are taken with wdio 4.9.10, though we have also tried with wdio 5 and it also uses a big amount of memory due to the multiprocess (little bit less than wdio 4).

For capabilities we are testing in chrome latest, and we have been able to test with 35 parallel browsers in a 4GB machine. Using more leads to an OOM kill.

```
{
        browserName:   "chrome",
        chromeOptions: {
            args: [
                '--disable-dev-shm-usage',
                '--headless',
                '--window-size=411,731' // Pixel 2 size
            ]
        },
        maxInstances: 35,
    }
```

**Comment by mgrybyk** [comment]
Wdio v5 won't spawn more than 36 nodejs insurances even if you have 1000 tests

**Comment by alfonso-presa** [comment]
@mgrybyk I thought the amount of processes spawned depended on the amount of spec files and the max instances. Do you mean that this is restricted somehow in v5?

Chrome is started in a cloud grid provider so our process is not impacted by the amount of memory it requires :-).

**Comment by christian-bromann** [comment]
@alfonso-presa I am able to easily spin up 100 instance using Sauce headless without having much memory consumption in v5.

**Comment by mgrybyk** [comment]
So yeah, I additional context would be helpful

**Comment by alfonso-presa** [comment]
@christian-bromann I've quickly made a couple of tests with v5:

* With a simple mocha test the footprint is around 50MB. We could consider this the minimum footprint per wdio browser.
* With the `cucumber-boilerplate` it reaches up to 85-90 MB, which is almost as much as we have with v4 in our case (100 MB) and will make ~8,5GB for  100 of tests in parallel.

IMHO there's a big range of improvement in the way webdriver.io is performing paralelization in terms of memory consumption. I don't know if you agree in that being able to reuse the same process for as much browsers as configured (i.e. 5 browsers in the same process) would be great in order to save in memory and enable better scalability. On the other hand I understand there might be limitations for this in nodejs, in the way webdriver.io is designed or even in the tests runners. If so, unfortunately, this improvement might not be viable.

One limitation I see is with the sync strategy... ¿Is syncing locking the event loop in v5?

The question is if you think this is worth the effort and if is something doable at first glance. If so, I would be very happy to try to implement it and PR.

**Comment by mgrybyk** [comment]
@alfonso-presa just checked it locally.
I have 20 chrome browser in capabilities.
Main thread consumes 33 Mb
Each worker thread consumes 150 Mb.

So, you are right, I think.

![image](https://user-images.githubusercontent.com/25589559/61714521-574ffd80-ad5b-11e9-8b92-f51798a10226.png)

**Comment by mgrybyk** [comment]
Running 100 browsers crushes my laptop
![image](https://user-images.githubusercontent.com/25589559/61714945-494eac80-ad5c-11e9-8e65-04677a61ec04.png)

**Comment by alfonso-presa** [comment]
@mgrybyk Thanks for looking at it. Those look like big tests also :-).

I'm thinking that also worker_threads might be a great way to sort this. May be this could be done with another runner module or a configurable option for the `local-runner`.... What do you think?

**Comment by christian-bromann** [comment]
> What do you think?

Depends on what you are proposing exactly.

**Comment by alfonso-presa** [comment]
@christian-bromann I mean trying to change (or provide a different implementation) the local-runner so that instead of forking it starts a [worker_thread](https://nodejs.org/docs/latest-v10.x/api/worker_threads.html). I've already done some testing and seems that also the `runner` and the `fibers` are impacted by this. I'm specially having a bad time with `fibers` trying to make it [context aware](https://nodejs.org/api/addons.html#addons_context_aware_addons) by using `NODE_MODULE_INITIALIZER`. I'm getting a `segfault` and I think it's because it has global variables defined that should be inside the context...

**Comment by alfonso-presa** [comment]
Looks like this is indeed going to require a lot of work :-) https://github.com/laverdet/node-fibers/issues/411

**Comment by mgrybyk** [comment]
Well, I'd like to start first with memory cleanup, then, if needed and possible move most memory consuming parts to services and only after think what can be done further.

After some investigation I found out the most memory is used by transpilers!
In my case, with mocha + typescript, it is 80% 

I want to check if it's possible to unload transpilers at some point or/and to load them once and share across workers.

What do you guys think?

**Comment by wswebcreation** [comment]
@mgrybyk 

> After some investigation I found out the most memory is used by transpilers!
In my case, with mocha + typescript, it is 80%

Are you transpiling on-the-fly or transpile all code before you run and then run all instances?

**Comment by mgrybyk** [comment]
On the fly

**Comment by alfonso-presa** [comment]
@mgrybyk That would be great. Any memory clean up will be very welcome :-).

In our case we are not using any transpilers and we are reaching up to 100MB per browser, so in order to solve this issue and save "orders of magnitude" in terms of memory consumption, a clean up will not be enough I think...

**Comment by mgrybyk** [comment]
That's gonna be the next step

**Comment by alfonso-presa** [comment]
An update to this issue.... I've been making some PoC with a modified runner and the footprint get's reduced a little bit by using 'worker_threads' (around 10MB), but memory consumption is still lineal with the number of browsers. Any further ideas?

**Comment by mgrybyk** [comment]
Not at this point, it's not something easy to do, sorry to say.

**Comment by christian-bromann** [comment]
Is there anything actionable to do here? @alfonso-presa we have done some memory optimisations since this thread was opened. Also keep in mind that running multiple browser in parallel definitely is memory costly. I've been running hundreds of jobs using Sauce Labs without any issues. Can someone summarise the issue here otherwise I will close it.

**Comment by alfonso-presa** [comment]
@christian-bromann I did some researching and worker_thread did hep a little bit, though memory was still an issue. I did also try to instantiate the driver in the same process several times ( > 50) and there was no memory issue at all, so the problem is in the runner itself. 

What I would expect from this issue is the runner to be able to bootstrap several tests in parallel without the need of bootstrapping several node processes. Although I understand the impact of this would be huge as for example there should not be global objects, which would make difficult the use of cucumber or mocha....

**Comment by christian-bromann** [comment]
@alfonso-presa are you running browser one the same machine as well? As I mentioned I am able to run hundreds of tests in parallel using Sauce Labs.

**Comment by alfonso-presa** [comment]
@christian-bromann I'm using saucelabs too If you have enough RAM (i.e. 16GB machine) you could probably run ~100 tests without problem (though it would take around 10GB). The problem arrives in our CI/CD environment where we have 4GB mem limit for our containerized slaves. We could rise that limit, but of course that does have a direct impact on infrastructure costs. If you think about it, those node processes are taking that memory just to copy the same code over and over again, as the state of each process will probably fit in less than 1MB...

**Comment by mgrybyk** [comment]
Is it possible (theoretically) to have test framework running as a service that would spawn dedicated instance (not a process) for every child process that is spawned by wdio runner?

Every test framework consumes some memory, let's say 30mb. So, if we need to run 10 tests in parallel it is 300 mb already.

Having only one test framework in memory should reduce memory usage dramatically. 

We can also think if there is anything else with high memory usage that can be moved to services.

**Comment by christian-bromann** [comment]
> Is it possible (theoretically) to have test framework running as a service

What do you mean by "running as a service"?

**Comment by mgrybyk** [comment]
Sorry, ignore my comment, it won't work like this.

**Comment by christian-bromann** [comment]
Do you have any numbers which parts consume which amount of memory?

**Comment by mgrybyk** [comment]
It greatly depends on test framework, services and reporters are used by user.
Every service/reporter should be investigated separately.

There is nothing we can do with test frameworks except of raising issues/PRs for them.
But there is something we can do with our services/reporters and dependencies we use.

However there are something can be done.

big size of some packages
- loading config increases memory usage by additional 5-7 Mb (investigating why)
`this.configParser.addConfigFile(configFile)`
-  `request` (because of multiple sub packages), this one for you @christian-bromann 
- `wdio-spec-reporter` - table (can we find an alternative for that?)
- `selenium-standalone` (urijs and some others). It shouldn't be loaded at all, see below.

Here is a big question why do we load services in child processes at all? It increases memory consumption for me for ~20Mb! May we chat on this one @christian-bromann ? If it's really needed we can have two types of services, some for just main process, others for child processes.

**Comment by mgrybyk** [comment]
@alfonso-presa what services/reporters are you using with wdio?

**Comment by christian-bromann** [comment]
> Here is a big question why do we load services in child processes at all?

Because they need to run certain functionality during the test process in hooks. Otherwise we would need to build logic that sends the data from the hook to the main process, computes them and sends it back to the child process.

> * `request` (because of multiple sub packages), this one for you @christian-bromann #3426

I am on it next week.

**Comment by mgrybyk** [comment]
@christian-bromann well at least it's not true for selenium-standalone, chromedriver, appium.

let's say no if we don't need service in child process and yes if we need.

chromedriver - no
sauce - ?
testingbot - ?
selenium-standalone - no
devtools - **yes**
applitools - **yes**
browserstack - ?
appium - no
firefox-profile - ?
zafira-listener - ?
reportportal - ?
crossbrowsertesting - ?
docker - ?

We avoid starting some services for child processes.

**Comment by christian-bromann** [comment]
> sauce	? (no)
> testingbot	? (no)
> browserstack	? (no)
> crossbrowsertesting	? (no)

We all need them as have to update sessions right after the session was closed.

Usually if a service has no worker part (e.g. Chromedriver service that only uses `onPrepare` and `onComplete` hooks) we just create an instance of an empty class `export default class AppiumService {}` that should not cost much memory. However I don't know if requiring the package adds to it because of storing files in the require cache. We could check in the launcher if the service provides a worker part and remove it if not.

**Comment by mgrybyk** [comment]
also these reporters consumes some memory:
- junit (because of junit-report-builder) 
- sumologic (because of request)

**Comment by mgrybyk** [comment]
> we just create an instance

@christian-bromann memory consumption happens not because we do something but because whole package and its sub modules are loaded. 
selenium-standalone consumes more than 10mb per worker for example

**Comment by christian-bromann** [comment]
> selenium-standalone consumes more than 10mb per worker for example

Interesting! Let's kick these services out of the service list if they don't have a service part for the launcher.

**Comment by mgrybyk** [comment]
yeah, let's start with 2 services (appium, selenium-standalone) first.
Have to double check everything first, will do somewhere next week

**Comment by mgrybyk** [comment]
I've also noticed that moving imports in wdio.conf from the root scope to `onPrepare` hook also improves memory usage.
Example:
```
// removed the import
import * as rimraf from "rimraf"

// moved to onPrepare hook
onPrepare: [function (config, capabilities) {
    const allureResultsDir = "./report/allure-results"
    require("rimraf").sync(allureResultsDir)
}],
```

**Comment by mgrybyk** [comment]
@alfonso-presa what services/reporters do you use with wdio?

If you use selenium-standalone and spec reporter you'll get -15Mb per worker instance soon.

Additionally if you can move some imports to `onPrepare` hook you might get some free Mb as well. It was 7 Mb for me.

I know it's not something you've expected but let's start with something

**Comment by alfonso-presa** [comment]
@mgrybyk Those sound like great improvements! Thank you!

We use selenium-standalone (but just for local, obviously not for wide scale executions) and spec reporter indeed. Also we use Sauce service and Appium (only local). Each Mb counts for us, so thank you very much for the effort :-).

**Comment by mgrybyk** [comment]
@alfonso-presa what setup you have when you need to run lots of tests in parallel?

**Comment by alfonso-presa** [comment]
@mgrybyk We run cucumber tests with spec reporter and sauce service, and we are also using the allure reporter.

**Comment by mgrybyk** [comment]
@alfonso-presa here you go https://github.com/webdriverio/webdriverio/pull/4624 and https://github.com/webdriverio/webdriverio/pull/4623

**Comment by mgrybyk** [comment]
Sauce package uses `request` as well and have to be replaced with something.

**Comment by christian-bromann** [comment]
@alfonso-presa @mgrybyk is the origin of these 150mb per worker the memory of Node.js alone or includes that the worker? Note that in WebDriver every session is an independent browser instance and Chrome alone (even running in headless) can consume a significant amount of memory (given that it is not just another tab).

**Comment by joscha** [comment]
> After some investigation I found out the most memory is used by transpilers!
> In my case, with mocha + typescript, it is 80%
> 
> I want to check if it's possible to unload transpilers at some point or/and to load them once and share across workers.

partially related: https://github.com/webdriverio/webdriverio/issues/4949

**Comment by christian-bromann** [comment]
Since the creation of this issue we have done various improvements on the performance. There are sure many more things to do but I would prefer to create new tickets for this that describe particular problems. Thank you everyone who were involved in this discussion.
